File: mobile/features/fragments/editorDocument.ts
import type { EditorBlock, EditorBlockType, EditorDocument, EditorMark, MediaAsset, FragmentAiPatch } from '@/types/fragment';

function createId(prefix: string): string {
  /** 中文注释：在客户端生成轻量块 ID，避免依赖服务端临时 ID。 */
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
}

export function emptyEditorDocument(): EditorDocument {
  /** 中文注释：返回空正文文档，统一用于新建和无内容回退。 */
  return { type: 'doc', blocks: [] };
}

export function buildDocumentFromText(text: string): EditorDocument {
  /** 中文注释：把普通文本包装成单段正文文档。 */
  const normalized = text.trim();
  if (!normalized) return emptyEditorDocument();
  return {
    type: 'doc',
    blocks: [
      {
        id: createId('block'),
        type: 'paragraph',
        children: [{ text: normalized, marks: [] }],
      },
    ],
  };
}

export function normalizeEditorDocument(document: EditorDocument | null | undefined): EditorDocument {
  /** 中文注释：规整服务端或本地文档，保证渲染期字段齐全。 */
  if (!document || document.type !== 'doc' || !Array.isArray(document.blocks)) {
    return emptyEditorDocument();
  }
  return {
    type: 'doc',
    blocks: document.blocks.map((block) => ({
      ...block,
      id: block.id || createId('block'),
      children: Array.isArray(block.children) ? block.children.map((child) => ({
        text: child.text ?? '',
        marks: Array.isArray(child.marks) ? child.marks.filter((mark): mark is EditorMark => mark === 'bold' || mark === 'italic') : [],
      })) : [],
    })),
  };
}

export function extractPlainText(document: EditorDocument | null | undefined): string {
  /** 中文注释：提取正文纯文本，供列表预览和状态判断复用。 */
  return normalizeEditorDocument(document).blocks
    .map((block) => {
      if (block.type === 'image') {
        return block.alt?.trim() || '';
      }
      return block.children.map((child) => child.text).join('').trim();
    })
    .filter(Boolean)
    .join('\n')
    .trim();
}

export function updateBlockText(document: EditorDocument, blockId: string, text: string): EditorDocument {
  /** 中文注释：更新指定文本块内容，保留现有样式标记。 */
  return {
    type: 'doc',
    blocks: normalizeEditorDocument(document).blocks.map((block) => {
      if (block.id !== blockId || block.type === 'image') return block;
      const firstChild = block.children[0] ?? { text: '', marks: [] };
      return {
        ...block,
        children: [{ text, marks: firstChild.marks }, ...block.children.slice(1).map((child) => ({ ...child, text: '' }))],
      };
    }),
  };
}

export function toggleBlockMark(document: EditorDocument, blockId: string, mark: EditorMark): EditorDocument {
  /** 中文注释：按整块切换粗体或斜体，保持首版交互简单。 */
  return {
    type: 'doc',
    blocks: normalizeEditorDocument(document).blocks.map((block) => {
      if (block.id !== blockId || block.type === 'image') return block;
      const firstChild = block.children[0] ?? { text: '', marks: [] };
      const marks = firstChild.marks.includes(mark)
        ? firstChild.marks.filter((item) => item !== mark)
        : [...firstChild.marks, mark];
      return {
        ...block,
        children: [{ text: firstChild.text, marks }, ...block.children.slice(1)],
      };
    }),
  };
}

export function setBlockType(document: EditorDocument, blockId: string, type: Exclude<EditorBlockType, 'image'>): EditorDocument {
  /** 中文注释：切换块级类型，用于标题、引用和列表。 */
  return {
    type: 'doc',
    blocks: normalizeEditorDocument(document).blocks.map((block) => block.id === blockId && block.type !== 'image' ? { ...block, type } : block),
  };
}

export function insertParagraphAfter(document: EditorDocument, blockId?: string | null): EditorDocument {
  /** 中文注释：在当前块后插入一个空段落，便于继续编辑。 */
  const nextBlock: EditorBlock = {
    id: createId('block'),
    type: 'paragraph',
    children: [{ text: '', marks: [] }],
  };
  const normalized = normalizeEditorDocument(document);
  if (!blockId) {
    return { type: 'doc', blocks: [...normalized.blocks, nextBlock] };
  }
  const nextBlocks: EditorBlock[] = [];
  let inserted = false;
  normalized.blocks.forEach((block) => {
    nextBlocks.push(block);
    if (block.id === blockId) {
      nextBlocks.push(nextBlock);
      inserted = true;
    }
  });
  if (!inserted) nextBlocks.push(nextBlock);
  return { type: 'doc', blocks: nextBlocks };
}

export function removeBlock(document: EditorDocument, blockId: string): EditorDocument {
  /** 中文注释：删除指定块，避免正文中残留无效空块。 */
  return {
    type: 'doc',
    blocks: normalizeEditorDocument(document).blocks.filter((block) => block.id !== blockId),
  };
}

export function insertImageBlock(document: EditorDocument, asset: MediaAsset, afterBlockId?: string | null): EditorDocument {
  /** 中文注释：把上传后的图片素材插入正文。 */
  const imageBlock: EditorBlock = {
    id: createId('image'),
    type: 'image',
    asset_id: asset.id,
    url: asset.file_url ?? null,
    width: asset.width ?? null,
    height: asset.height ?? null,
    alt: asset.original_filename,
    children: [],
  };
  const normalized = normalizeEditorDocument(document);
  const nextBlocks: EditorBlock[] = [];
  let inserted = false;
  normalized.blocks.forEach((block) => {
    nextBlocks.push(block);
    if (block.id === afterBlockId) {
      nextBlocks.push(imageBlock);
      inserted = true;
    }
  });
  if (!inserted) nextBlocks.push(imageBlock);
  return { type: 'doc', blocks: nextBlocks };
}

export function collectDocumentAssetIds(document: EditorDocument): string[] {
  /** 中文注释：从正文图片节点中提取素材引用，便于和接口同步。 */
  const ids: string[] = [];
  normalizeEditorDocument(document).blocks.forEach((block) => {
    if (block.type !== 'image') return;
    const assetId = block.asset_id?.trim();
    if (assetId && !ids.includes(assetId)) ids.push(assetId);
  });
  return ids;
}

export function applyAiPatch(document: EditorDocument, patch: FragmentAiPatch): EditorDocument {
  /** 中文注释：把后端 AI patch 应用到本地文档。 */
  const normalized = normalizeEditorDocument(document);
  if (patch.op === 'prepend_heading' && patch.block) {
    return { type: 'doc', blocks: [patch.block, ...normalized.blocks] };
  }
  if (patch.op === 'insert_after_selection') {
    const blocks = patch.blocks ?? [];
    return {
      type: 'doc',
      blocks: insertAfterTarget(normalized.blocks, patch.target_block_id ?? null, blocks),
    };
  }
  if (patch.op === 'replace_selection') {
    return {
      type: 'doc',
      blocks: normalized.blocks.map((block) => {
        if (block.id !== patch.target_block_id || block.type === 'image') return block;
        const firstChild = block.children[0] ?? { text: '', marks: [] };
        return {
          ...block,
          children: [{ text: patch.replacement_text ?? '', marks: firstChild.marks }],
        };
      }),
    };
  }
  return normalized;
}

function insertAfterTarget(blocks: EditorBlock[], targetBlockId: string | null, insertion: EditorBlock[]): EditorBlock[] {
  /** 中文注释：在指定块后插入一组新块，找不到时追加到末尾。 */
  if (!targetBlockId) return [...blocks, ...insertion];
  const nextBlocks: EditorBlock[] = [];
  let inserted = false;
  blocks.forEach((block) => {
    nextBlocks.push(block);
    if (block.id === targetBlockId) {
      nextBlocks.push(...insertion);
      inserted = true;
    }
  });
  if (!inserted) nextBlocks.push(...insertion);
  return nextBlocks;
}

File: backend/modules/shared/editor_document.py
from __future__ import annotations

from copy import deepcopy
from typing import Any

from core.exceptions import ValidationError


ALLOWED_BLOCK_TYPES = {"paragraph", "heading", "blockquote", "bullet_list", "ordered_list", "image"}
ALLOWED_MARKS = {"bold", "italic"}
EMPTY_EDITOR_DOCUMENT: dict[str, Any] = {"type": "doc", "blocks": []}


def empty_editor_document() -> dict[str, Any]:
    """返回稳定的空文档结构，避免直接复用可变默认值。"""
    return deepcopy(EMPTY_EDITOR_DOCUMENT)


def build_document_from_text(text: str, *, block_type: str = "paragraph") -> dict[str, Any]:
    """把纯文本快速包装成最小可用的富文本文档。"""
    normalized = text.strip()
    if not normalized:
        return empty_editor_document()
    return {
        "type": "doc",
        "blocks": [
            {
                "id": "block-1",
                "type": block_type,
                "children": [
                    {
                        "text": normalized,
                        "marks": [],
                    }
                ],
            }
        ],
    }


def normalize_editor_document(document: Any) -> dict[str, Any]:
    """校验并规整富文本文档结构，拒绝未知块和非法 mark。"""
    if document is None:
        return empty_editor_document()
    if not isinstance(document, dict):
        raise ValidationError(message="正文文档格式无效", field_errors={"editor_document": "必须是对象"})
    if document.get("type") != "doc":
        raise ValidationError(message="正文文档格式无效", field_errors={"editor_document.type": "必须是 doc"})
    raw_blocks = document.get("blocks")
    if raw_blocks is None:
        return empty_editor_document()
    if not isinstance(raw_blocks, list):
        raise ValidationError(message="正文文档格式无效", field_errors={"editor_document.blocks": "必须是数组"})

    normalized_blocks: list[dict[str, Any]] = []
    for index, raw_block in enumerate(raw_blocks):
        if not isinstance(raw_block, dict):
            raise ValidationError(message="正文块格式无效", field_errors={f"editor_document.blocks.{index}": "必须是对象"})
        block_type = str(raw_block.get("type") or "").strip()
        if block_type not in ALLOWED_BLOCK_TYPES:
            raise ValidationError(
                message="正文块类型无效",
                field_errors={f"editor_document.blocks.{index}.type": f"仅支持 {', '.join(sorted(ALLOWED_BLOCK_TYPES))}"},
            )
        block_id = str(raw_block.get("id") or f"block-{index + 1}").strip() or f"block-{index + 1}"
        if block_type == "image":
            normalized_blocks.append(
                {
                    "id": block_id,
                    "type": "image",
                    "asset_id": str(raw_block.get("asset_id") or "").strip() or None,
                    "url": str(raw_block.get("url") or "").strip() or None,
                    "width": _coerce_int(raw_block.get("width")),
                    "height": _coerce_int(raw_block.get("height")),
                    "alt": str(raw_block.get("alt") or "").strip() or None,
                }
            )
            continue

        children = raw_block.get("children")
        if not isinstance(children, list):
            raise ValidationError(
                message="正文块格式无效",
                field_errors={f"editor_document.blocks.{index}.children": "必须是数组"},
            )
        normalized_children: list[dict[str, Any]] = []
        for child_index, raw_child in enumerate(children):
            if not isinstance(raw_child, dict):
                raise ValidationError(
                    message="正文文本片段格式无效",
                    field_errors={f"editor_document.blocks.{index}.children.{child_index}": "必须是对象"},
                )
            text = str(raw_child.get("text") or "")
            raw_marks = raw_child.get("marks") or []
            if not isinstance(raw_marks, list):
                raise ValidationError(
                    message="正文文本样式格式无效",
                    field_errors={f"editor_document.blocks.{index}.children.{child_index}.marks": "必须是数组"},
                )
            marks = []
            for raw_mark in raw_marks:
                normalized_mark = str(raw_mark or "").strip()
                if normalized_mark not in ALLOWED_MARKS:
                    raise ValidationError(
                        message="正文文本样式无效",
                        field_errors={
                            f"editor_document.blocks.{index}.children.{child_index}.marks": f"仅支持 {', '.join(sorted(ALLOWED_MARKS))}"
                        },
                    )
                if normalized_mark not in marks:
                    marks.append(normalized_mark)
            normalized_children.append({"text": text, "marks": marks})
        normalized_blocks.append({"id": block_id, "type": block_type, "children": normalized_children})
    return {"type": "doc", "blocks": normalized_blocks}


def extract_plain_text_from_document(document: dict[str, Any] | None) -> str:
    """从富文本文档中提取纯文本快照，用于检索、摘要和生成。"""
    if not document or not isinstance(document, dict):
        return ""
    parts: list[str] = []
    for block in document.get("blocks") or []:
        if not isinstance(block, dict):
            continue
        if block.get("type") == "image":
            alt = str(block.get("alt") or "").strip()
            if alt:
                parts.append(alt)
            continue
        children = block.get("children") or []
        text = "".join(str(item.get("text") or "") for item in children if isinstance(item, dict)).strip()
        if text:
            parts.append(text)
    return "\n".join(parts).strip()


def collect_asset_ids_from_document(document: dict[str, Any] | None) -> list[str]:
    """收集正文内嵌图片节点引用的素材 ID。"""
    if not document or not isinstance(document, dict):
        return []
    asset_ids: list[str] = []
    for block in document.get("blocks") or []:
        if not isinstance(block, dict) or block.get("type") != "image":
            continue
        asset_id = str(block.get("asset_id") or "").strip()
        if asset_id and asset_id not in asset_ids:
            asset_ids.append(asset_id)
    return asset_ids


def render_document_as_markdown(document: dict[str, Any] | None) -> str:
    """把富文本文档单向渲染为 Markdown 导出文本。"""
    if not document or not isinstance(document, dict):
        return ""
    rendered_blocks: list[str] = []
    for block in document.get("blocks") or []:
        if not isinstance(block, dict):
            continue
        block_type = str(block.get("type") or "").strip()
        if block_type == "image":
            url = str(block.get("url") or "").strip()
            alt = str(block.get("alt") or "").strip()
            if url:
                rendered_blocks.append(f"![{alt}]({url})")
            continue
        text = "".join(_render_child_markdown(item) for item in block.get("children") or [] if isinstance(item, dict)).strip()
        if not text:
            continue
        if block_type == "heading":
            rendered_blocks.append(f"# {text}")
        elif block_type == "blockquote":
            rendered_blocks.append(f"> {text}")
        elif block_type == "bullet_list":
            rendered_blocks.append(f"- {text}")
        elif block_type == "ordered_list":
            rendered_blocks.append(f"1. {text}")
        else:
            rendered_blocks.append(text)
    return "\n\n".join(rendered_blocks).strip()


def apply_ai_patch(document: dict[str, Any], patch: dict[str, Any]) -> dict[str, Any]:
    """按后端返回的 patch 更新文档，保持移动端和服务端语义一致。"""
    normalized = normalize_editor_document(document)
    operation = str(patch.get("op") or "").strip()
    blocks = normalized.get("blocks") or []
    if operation == "prepend_heading":
        heading = normalize_editor_document({"type": "doc", "blocks": [patch.get("block")]}).get("blocks", [])
        return {"type": "doc", "blocks": heading + blocks}
    if operation == "insert_after_selection":
        insertion = normalize_editor_document({"type": "doc", "blocks": patch.get("blocks") or []}).get("blocks", [])
        target_block_id = str(patch.get("target_block_id") or "").strip()
        if not target_block_id:
            return {"type": "doc", "blocks": blocks + insertion}
        next_blocks: list[dict[str, Any]] = []
        inserted = False
        for block in blocks:
            next_blocks.append(block)
            if block.get("id") == target_block_id:
                next_blocks.extend(insertion)
                inserted = True
        if not inserted:
            next_blocks.extend(insertion)
        return {"type": "doc", "blocks": next_blocks}
    if operation == "replace_selection":
        target_block_id = str(patch.get("target_block_id") or "").strip()
        replacement_text = str(patch.get("replacement_text") or "")
        next_blocks: list[dict[str, Any]] = []
        for block in blocks:
            if block.get("id") != target_block_id:
                next_blocks.append(block)
                continue
            updated_block = deepcopy(block)
            if updated_block.get("type") == "image":
                next_blocks.append(updated_block)
                continue
            children = updated_block.get("children") or []
            if children:
                children[0]["text"] = replacement_text
                for index in range(1, len(children)):
                    children[index]["text"] = ""
            else:
                updated_block["children"] = [{"text": replacement_text, "marks": []}]
            next_blocks.append(updated_block)
        return {"type": "doc", "blocks": next_blocks}
    raise ValidationError(message="AI patch 无效", field_errors={"patch.op": "不支持的操作"})


def _render_child_markdown(child: dict[str, Any]) -> str:
    """把带 mark 的文本片段渲染为 Markdown 内联片段。"""
    text = str(child.get("text") or "")
    marks = child.get("marks") or []
    if "italic" in marks:
        text = f"*{text}*"
    if "bold" in marks:
        text = f"**{text}**"
    return text


def _coerce_int(value: Any) -> int | None:
    """把可选尺寸字段规整为整数。"""
    if value in (None, ""):
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None
